Rethink code navigation with CodeMirror token selection

olafurpg opened the pull request:
Previously, the token cursor feature flag used HTML href links for go to definition. This commit changes the behavior of the feature flag to use CodeMirror selections instead along with several other big changes:

* Cmd+Click for Go to definition
* Enter for Go to definition
* Space for Show popover
* Up/Down/Right/Left arrow to move the selection to the next token

Demo https://www.loom.com/share/a0fa53badd4d419081dcfd1a2fbc7e4e

Changes since the demo was recorded

* The hover popover looks the same as in the main branch
* No more context menu
* Improved accessibility
* Improved history handling

Please give it a try and let us know what you think.

TODOs before this PR is ready for review:

- [x] Underline token when definition is available and meta key is being held
- [x] Update selection when moving history inside current file
- [x] Handle multiple definitions by linking to ref panel
- [x] Document highlights
- [x] Use ctrl key on linux and windows
- [x] Highlight hovered token
- [x] Only show pointer when over an interactive token
- [x] Separate tooltip for hover and definition
- [x] Allow cmd+click on any token including non-occurrences
- [x] Move feature flag under experimental
- [x] Fix position of "No definition found" tooltip (show at cursor)
- [x] Fix stack trace from Thorsten
- [x] Fix stack trace from Noah. Steps: select token, space for popover, click on right margin.
- [x] Render link for "right-click > open in new tab" and middle-click
- [x] Support long-press goto-def
- [x] Support comprehensive navigation in search-based intel for languages with low-quality highlighting data
- [x] Document new goto-def triggers in hover popover
- [x] Trigger find-ref on goto-def at the definition (save vscode and intellij)
- [x] Avoid pushing cycles into history (repeated goto-def from same location)
- [x] Fix failing test cases


Improvements for the future: see https://github.com/sourcegraph/sourcegraph/issues/45006


## Test plan

To test it manually with a local build. First start the server. Note that it's not possible to use `web-standalone`.

```
sg start 
```

Next, update setting `"experimentalFeatures": {"codeNavigation": "selection-driven"}}`.

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

## App preview:

- [Web](https://sg-web-olafurpg-definition-button.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.01% (+0.31 kb) | -0.03% (-4.76 kb) | -0.04% (-5.07 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/olafurpg/token-selection/compare-report.html;tab=live_object) for a full comparison between the commits 93cae6c and 5ee629d or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

olafurpg commented:
I prematurely marked this PR as ready for review. I'm very happy with the functionality, but I'm planning to self-review the diff on Monday and clean up the code before tagging people for review.

olafurpg commented on client/web/src/repo/blob/BlobPage.tsx:
```diff
@@ -561,12 +573,13 @@ export const BlobPage: React.FunctionComponent<React.PropsWithChildren<BlobPageP
                         disableStatusBar={!window.context.enableLegacyExtensions}
                         disableDecorations={false}
-                        role="region"
+                        role="main"
```
@limitedmage Can you advice on the most appropriate role for the blob view? The "region" role did not work well with hjlk navigation.

olafurpg replied:
You're totally right that `main` is inappropriate. I've reverted this change so that it's `region` again. Selection driven navigation will need more iterations to get accessibility right. When I tried hjkl navigation with the `region` role then VoiceOver didn't speak out the selection.

limitedmage replied:
I think `region` is fine. Does it show up in the Landmarks menu? (VoiceOver+U, use left/right arrows to go between menus).

olafurpg commented:
@fkling I've cleaned up this PR to the point where I'm decently happy with the code organization. The biggest missing piece is e2e tests for the new code paths, but we don't have e2e tests for the 'tokens-as-links' setting either.

All of the new functionality should be hidden behind the feature flag so I'd like to get this merged to collect more internal feedback before investing more time in e2e tests.

lrhacker commented:
This is awesome - I will make time to take a look at it tomorrow! (cc @sourcegraph/code-exploration for visibility)

@olafurpg Would you be able to compare the work you've done against the items listed in [this tracking issue](https://github.com/sourcegraph/sourcegraph/issues/43176) and note which issues you think it should close in the PR?

felixfbecker commented:
I'm getting when trying this out, is there something I'm doing wrong?

> invalid settings: experimentalFeatures: Additional property codeNavigation is not allowed

<img width="848" alt="image" src="https://user-images.githubusercontent.com/10532611/204407681-6380ea69-6382-4bdd-adba-54d51b14741a.png">

olafurpg commented:
@felixfbecker Have you tried with `sg start`? This feature flag is not available with web-standalone.

olafurpg commented:
@lrhacker This PR enables keyboard navigation but I don't think the solution is fully accessible as is. When using VoiceOver, the currently selected token is read out but you have to use hjkl navigation instead of up/down/right/left. To achieve this, I made a small change to use the `main` role for the blob component (instead of `region`) and adding the `aria-current=true` attribute to the selected token. We can fairly easily customize the HTML structure to add any other roles or aria attributes, but I don't think we can use links for the reasons mentioned here https://docs.google.com/document/d/1cXZEXfui1ylXBfNoB5iK7sUYhbNm51Y7wQMW36RScTA/edit#

lrhacker commented:
A few different chunks of feedback and comments so far. I like @taras-yemets's table but this might be a little verbose for columns (sorry 😅).

---

From that tracking issue I linked, I do think this satisfies #43175, in that its goal is to include up/down/left/right keyboard navigation. The primary difference from the acceptance criteria I see is that up/down navigation now focuses on a single token, whereas what @umpox had prototyped originally was selecting just line to start with (and left/right navigation would start token selection).

One question: in your Loom it appears that the current line is highlighted along with the token, but I'm not seeing that locally - is that an expected change? (I personally prefer to see the line highlight so it's easier to see where I am.)

<img width="592" alt="image" src="https://user-images.githubusercontent.com/10523985/204648219-2e47c52d-f9d2-4bf8-bf88-0b50cb6b3d2f.png">

---

For #42421 from the tracking issue, while I wouldn't expect this PR itself to include tab-driven navigation, that is something I find myself wanting. Or, maybe more specifically, I want to make sure that pressing `tab` doesn't put me in an unexpected navigation situation. Currently, if `tab` is typed while in the blob view, it will switch focus to the actions panel, and then the keyboard navigation/controls for the blob view do not work. If you tried to hit `space` at this point to close an open intel popover, it would instead trigger a click on the next element that's focused (in my case, the "View on GitHub" link).

---

The [RFC](https://docs.google.com/document/d/1_r8aikiE-zwn0ZkWmOilynNVA3HfmF9Mglgg6ZEkkls/edit#) that @taras-yemets linked above will also be something to consider and be a good place to discuss what specific keyboard interactions we want to settle on. Not that we can't differ from what's in the RFC today, but it will be good to make sure we're all in agreement with what is expected (e.g. should `enter` immediately go to definition like in this PR, or just open the intel popover, as suggested in the RFC and #42421).

Side note: one specific interaction specified in the RFC that I found myself missing here, though, is using `esc` close the intel popover. 😄 

---

In terms of accessibility, I still need to do more testing, but I think we'll need to keep iterating on interactions with VoiceOver (or, more generally, all screen readers) over time. At the very least, it's not always the simplest problem to solve on the first pass.

When I did a very quick test, using the actual VoiceOver controls for navigation (e.g., Ctrl + Opt + Right arrow) does not selecting things as expected (the entire row of text is selected/read). I still need to do more research on standards around supporting specific screen reader controls, versus having other keyboard-based navigation available (e.g. h/j/k/l) that works in tandem with the reader to announce something meaningful. @limitedmage do you happen to have any resources on this?

That being said, there are almost certainly some things we can do using ARIA roles to better describe the tokens. The first thing that comes to mind would be to use something like the [`aria-expanded`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded) (or similar) role on each token to indicate it has a menu it can open (the intel popover). That might not be _exactly_ the correct role in this case, but the general idea is that we should be able to find a way to mark them as interactive and what they are intended to do.

---

Lastly, like @taras-yemets mentioned, the intel popover does not close when you begin navigating again. I think the preferred interaction for now would actually be to close it when you start navigating again, because it easily has the potential to cover/hide where the selection has moved to, like so:

<img width="365" alt="image" src="https://user-images.githubusercontent.com/10523985/204644385-452e0744-52a1-4ee2-9084-2cf71ebb5586.png">

olafurpg commented on client/web/src/repo/blob/codemirror/token-selection/modifier-click.ts:
```diff
@@ -0,0 +1,63 @@
+        }
+        public onKeyDown = (event: KeyboardEvent): void => {
+            if (isModifierKey(event) && !this.view.state.field(isModifierKeyHeld)) {
+                this.view.contentDOM.classList.add('cm-token-selection-clickable')
```
Nice! I didn't know this was possible, the code is much cleaner with this change.

olafurpg commented:
@taras-yemets @lrhacker thank you so much for the helpful review 🙏🏻 

> Side note: one specific interaction specified in the RFC that I found myself missing here, though, is using esc close the intel popover. 😄

I've fixed this, was missing a `closeHover()` in the `Escape` handler.

> Lastly, like @taras-yemets mentioned, the intel popover does not close when you begin navigating again. I think the preferred interaction for now would actually be to close it when you start navigating again, because it easily has the potential to cover/hide where the selection has moved to, like so:

The popover always disappears now when the editor selection changes (regardless if you use the mouse or arrow keys. Thank you for spotting this bug!

> If we open the popup for a selected token and then navigate to another one and want to open a code-intel popup for it we need to press 'Space' two times: the first one to close the existing code-intel popup and the second one to open the one for the currently selected token

This is fixed with the above change to hide the popover when selection changes.

> One question: in your Loom it appears that the current line is highlighted along with the token, but I'm not seeing that locally - is that an expected change? (I personally prefer to see the line highlight so it's easier to see where I am.)

I've opened a "followup" issue with a checkbox for this feedback https://github.com/sourcegraph/sourcegraph/issues/45006  I removed `history.replace()` for the selected token it it added a noticeable delay (~200ms) to re-render the entire app based on the new URL. A possible compromise is to highlight the line without updating the URL.

>  should enter immediately go to definition like in this PR, or just open the intel popover,

We can go either way and can make the decision based on user research. I personally prefer supporting goto-def with a single "premium" shortcut like `Enter` (similar to how `Enter` navigates to a link).

> I got some unexpected code-intel popup content which I didn't get on dotcom. Not sure if it may be caused by my local Sourcegraph instance.

This is just an example of bad search-based code intel. When we implement tree-sitter highlighting for TypeScript then we can make `number` and `boolean` "built-in types" so that they don't have popovers by default.

> Some tokens (like the arrow on the screenshot) are not expected to have valuable code intel information. Maybe they shouldn't support 'Space' clicks for code-intel popups display. Hovering these tokens on dotcom doesn't trigger the code-intel popup.

Same issue as above, that will be fixed with a tree-sitter highlighter, which we're planning to fix once @tjdevries is done with scip-go.

> 'You are at the definition' and 'Loading' popups need some padding and brand styling

Good call! I've added this to the followup issue. The current tooltips are using the default style for codemirror tooltips and we can do anything we like, including rendering react components like we do for the codeintel popover (kudos to @fkling for cracking the nut on how to implement that!). https://github.com/sourcegraph/sourcegraph/issues/45006

olafurpg commented:
> For https://github.com/sourcegraph/sourcegraph/issues/42421 from the tracking issue, while I wouldn't expect this PR itself to include tab-driven navigation, that is something I find myself wanting. Or, maybe more specifically, I want to make sure that pressing tab doesn't put me in an unexpected navigation situation.

I've added this to the followup issue #45006. I originally implemented `Tab`/`Shift-Tab` support but removed it because I don't want to risk keyboard traps like the `Tab` override that was part of the original fuzzy finder https://www.w3.org/TR/WCAG21/#no-keyboard-trap

> At this point, the goal would just be to make sure we aren't intentionally doing anything in this PR that would make supporting tab-driven navigation prohibitively challenging in the future.

We have the full power of the CodeMirror API to implement whatever behavior we like for `Tab` and `Shift-Tab`. However, I don't recommend using HTML links for the reasons listed in https://docs.google.com/document/d/1cXZEXfui1ylXBfNoB5iK7sUYhbNm51Y7wQMW36RScTA/edit# I also personally like being able to safely click on the code without navigating somewhere else (Cmd+Click and `Enter` feel nice IMO).

olafurpg commented:
Is someone able to give this PR an approval so that we can merge it and collect more internal feedback? The functionality is behind a feature flag so we can always remove it in the future. I did my best to organize the code so that it's easy to remove by keeping the files in a separate `token-selection` directory.

lrhacker commented on client/web/src/repo/blob/codemirror/token-selection/keybindings.ts:
```diff
@@ -0,0 +1,163 @@
+}
+
+export const tokenSelectionKeyBindings = keybindings.flatMap(keybinding => {
+    const remappedKey = remappedKeyBindings?.[keybinding.key ?? '']
```
```suggestion
    const remappedKey = remappedKeyBindings[keybinding.key ?? '']
```

Can it just be this?

olafurpg replied:
This works 👍🏻 Thanks

lrhacker reviewed (approved):
No major concerns with the implementation itself at this point, just some minor feedback there. I also asked the rest of @sourcegraph/code-exploration-devs to make sure to take a look. I do think it will be good to get this in front of some more folks as an experimental feature to facilitate more feedback on the behavior (which could also possibly help inform some decisions in [RFC 345](https://docs.google.com/document/d/1_r8aikiE-zwn0ZkWmOilynNVA3HfmF9Mglgg6ZEkkls/edit#)).

As a side note, [I did add a comment](https://github.com/sourcegraph/sourcegraph/issues/45006#issuecomment-1334498005) to your follow-up issue that if we commit to taking this to GA, some more documentation will be really important.

olafurpg commented on client/client-api/src/contribution.ts:
```diff
@@ -81,7 +81,8 @@ export interface ActionContribution {
     commandArguments?: (TemplateExpression | number | boolean | null | object | any[])[]
 
-    /** The title that succinctly describes what this action does. */
+    /** The title that succinctly describes what this action does. The question
+     * mark '?' renders the MDI HelpCircleOutline icon. */
     title?: TemplateExpression
```
Great suggestion. Done.

olafurpg commented on client/web/src/repo/blob/BlobPage.tsx:
```diff
@@ -294,7 +306,7 @@ export const BlobPage: React.FunctionComponent<React.PropsWithChildren<BlobPageP
      */
     const stencil = useObservable(
```
Updated the comment 👍🏻

olafurpg commented on client/web/src/repo/blob/CodeMirrorBlob.tsx:
```diff
@@ -179,19 +186,23 @@ export const Blob: React.FunctionComponent<BlobProps> = props => {
-                      blobInfo,
-                      preloadGoToDefinition,
-                  })
+            enableSelectionDrivenCodeNavigation && codeintel
+                ? tokenSelectionExtension(codeintel, blobInfo, history)
                 : [],
+            !enableLinkDrivenCodeNavigation ? [] : tokensAsLinks({ history, blobInfo, preloadGoToDefinition }),
```
Agreed, fixed.

olafurpg commented on client/web/src/repo/blob/codemirror/occurrence-utils.ts:
```diff
@@ -0,0 +1,197 @@
+    SyntaxKind.IdentifierAttribute,
+])
+
+export const isInteractiveOccurrence = (occurence: Occurrence): boolean => {
+    if (!occurence.kind) {
+        return false
+    }
+
+    return INTERACTIVE_OCCURRENCE_KINDS.has(occurence.kind)
+}
```
Nice catch! This is such an easy typo to make 😅 

FWIW, the name "occurrence" comes from SCIP (https://sourcegraph.com/github.com/sourcegraph/scip/-/blob/scip.proto?L369), which is a name that is originally used in XCode's internal "index-while-building" persistence format. I'm not married to this name, but a lot of tempting alternatives like "token" are technically incorrect.

olafurpg commented:
@lrhacker Thank you so much for the review, great suggestions 🙏🏻

taras-yemets commented:
@olafurpg, I tried a few keyboard and screen reader scenarios and here's what I got.

1. Keyboard Tab/Shift+Tab navigation works good. I can navigate both code symbols and tree view items depending on where the focus is. 
 
https://user-images.githubusercontent.com/25318659/205272480-e1e8b3d8-9fe7-40cc-ae24-9b53ec8d0af0.mov


2. When using screen reader I can have code lines be read line by line but I don't have symbols keyboard navigation (⬆️ ⬇️ ⬅️ ➡️ ):
  - Control+Option+➡️ announces code line by line, but arrow keys navigate through the file tree
  - pressing Space on a token results in loading the selected file in the file tree (sec 42 on video below) 

https://user-images.githubusercontent.com/25318659/205272165-04160dd7-8f18-477c-9c87-91c54860a1e1.mov


3. When I focus the code view and then enable the screen reader I can navigate symbols with the arrow keys, but it behaves a bit unexpectedly:
  - I can't go to the next line with Control+Option+➡️ and get returned back to the current selection (sec 00-06 on the video below)
  - first symbol on the line is announced fully, but the next ones have only the last letter announced; when going backwards, punctuation signs but not symbols are announced (sec 10-30 on the video below)

https://user-images.githubusercontent.com/25318659/205274524-3146a9b8-efef-47dc-9d7b-4d6dd114ebff.mov

taras-yemets commented on client/web/src/repo/blob/codemirror/token-selection/facets.ts:
```diff
@@ -0,0 +1,30 @@
+export const textDocumentImplemenationSupport = Facet.define<boolean, boolean>({
+    combine: props => props[0],
+})
+export const uriFacet = Facet.define<string, string>({
+    combine: props => props[0],
+})
```
@olafurpg, I'm not sure I understand how this facet is used. I can see it being the part of `tokenSelectionExtension` but can't find other references to it.

olafurpg replied:
Good catch, this facet is no longer used after I replaced an old solution with a general purpose one.

olafurpg replied:
Deleted this facet.

taras-yemets commented on client/web/src/repo/blob/codemirror/token-selection/facets.ts:
```diff
@@ -0,0 +1,30 @@
+export const uriFacet = Facet.define<string, string>({
+    combine: props => props[0],
+})
+export const blobInfoFacet = Facet.define<BlobInfo, BlobInfo>({
+    combine: props => props[0],
+})
```
[`blobPropsFacet`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@5963a38026dbc3fe13dc482b46b4dfd2dad54a1e/-/blob/client/web/src/repo/blob/codemirror/index.ts?L5-11) enables extensions access to [`BlobProps`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@5963a38026dbc3fe13dc482b46b4dfd2dad54a1e/-/blob/client/web/src/repo/blob/Blob.tsx?L123) which already contain `BlobInfo`. Can we use `blobPropsFacet` instead of this one?

olafurpg replied:
Indeed. Removed!

taras-yemets commented on client/web/src/repo/blob/codemirror/token-selection/facets.ts:
```diff
@@ -0,0 +1,30 @@
+export const blobInfoFacet = Facet.define<BlobInfo, BlobInfo>({
+    combine: props => props[0],
+})
+export const codeintelFacet = Facet.define<Remote<FlatExtensionHostAPI>, Remote<FlatExtensionHostAPI>>({
```
```suggestion
export const codeIntelFacet = Facet.define<Remote<FlatExtensionHostAPI>, Remote<FlatExtensionHostAPI>>({
```
nit: camelCasing the variable name for consistency with the rest of the codebase

olafurpg replied:
Done.

taras-yemets commented on client/web/src/repo/blob/codemirror/token-selection/facets.ts:
```diff
@@ -0,0 +1,30 @@
+    // "go to definition" twice in a row from the same location.
+    previousURL?: string
+}
+export const historyFacet = Facet.define<H.History, H.History<HistoryState>>({
+    combine: props => props[0] as H.History<HistoryState>,
+})
```
[`blobPropsFacet`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@5963a38026dbc3fe13dc482b46b4dfd2dad54a1e/-/blob/client/web/src/repo/blob/codemirror/index.ts?L5-11) enables extensions access to [`BlobProps`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@5963a38026dbc3fe13dc482b46b4dfd2dad54a1e/-/blob/client/web/src/repo/blob/Blob.tsx?L119) which already contains `history`. Can we use `blobPropsFacet` instead of this one?

olafurpg replied:
Good catch. I removed this facet.

taras-yemets commented on client/web/src/repo/blob/codemirror/tooltips/CodeIntelTooltip.tsx:
```diff
@@ -0,0 +1,95 @@
+                    },
+                },
+            ]
+            if (hover.isPrecise && view.state.facet(textDocumentImplemenationSupport)) {
```
```suggestion
            if (hover.isPrecise && view.state.facet(blobPropsFacet).blobInfo.mode) {
```
I wonder if we can make use of existing `blobPropsFacet`, WDYT?

olafurpg commented:
@taras-yemets Thank you for the detailed review. Thanks to your suggestions, I was able to remove `facets.ts` altogether including `codeintelFacet` resulting in much cleaner code 😊 The token-selection extension now accepts no parameters since everything is already part of `blobPropsFacet`.

Can you please repost your accessibility comment in https://github.com/sourcegraph/sourcegraph/issues/45006 ? It's outside the scope of this PR to fix accessibility.
